feat: Avoid coloring moved lines

gerhardol opened the pull request:
Fixes #11843
Depends on #11862 

## Proposed changes

* feat: Avoid coloring moved lines

Use heuristics to select which lines that are candidates to
mark inlines for, avoid lines that Git has detected
as moved lines.
(the change is not complete, a few parts required from the second commit, to be squashed befor merge)

* feat: Only parse diff files once

Only parse the document for added/removed lines once, by using
the DiffLineInfo info parsed for the linenumbering.
Previously, the document was parsed many times every time the
highlighting (including searching and marking) was changed.
This simplifies the code.

## Test methodology <!-- How did you ensure quality? -->

Some kind of tests will be added for the gap change, the other visible change has updated tests.

## Merge strategy

Rebase

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:
```diff
@@ -156,6 +162,14 @@ public static DiffLinesInfo Analyze(TextEditorControl textEditor, bool isCombine
+            => textMarkers.Count > 0
+                && !MarkerColorMatch(textMarkers[0], meta.LineType)
+                && (textMarkers.Count <= 1 || !MarkerColorMatch(textMarkers[^1], meta.LineType) || text.AsSpan()[textMarkers[^1].Offset..textMarkers[^1].EndOffset].IsWhiteSpace())
+                && (textMarkers.Count <= 2 || !textMarkers[1..^1].All(i => MarkerColorMatch(i, meta.LineType)));
```
```suggestion
                && (textMarkers.Count <= 2 || !textMarkers[1..^1].All(m => MarkerColorMatch(m, meta.LineType)));
```

mstv commented on src/app/GitUI/Editor/Diff/DiffLinesInfo.cs:
```diff
@@ -1,4 +1,4 @@
-﻿namespace GitUI.Editor.Diff;
+namespace GitUI.Editor.Diff;
```
The BOM got lost. Please revert.

mstv commented on src/app/GitUI/Editor/Diff/GrepHighlightService.cs:
```diff
@@ -5,29 +5,30 @@
+        lineNumbersControl.DisplayLineNum(_diffLinesInfo, showLeftColumn: false);
+    }
+
+    public override void AddTextHighlighting(IDocument document)
+        => document.MarkerStrategy.AddMarkers(_textMarkers);
```
❔ Move to a base class?

gerhardol replied:
That would be TextHighlightService, but there are no textMerkers there.
So overridden in Diff (base), Grep and Difftastic.

mstv reviewed (commented):
partial review of first commit:

* test `CanGetAddedMovedDiffInfo` needs to be adapted

RussKie commented on src/app/GitUI/Editor/Diff/Segment.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace GitUI.Editor.Diff;
+
+public struct Segment : ISegment
```
Why do we need the interface?

gerhardol replied:
It is a TextEditor interface

gerhardol replied:
There is one implementation in externals\ICSharpCode.TextEditor\Project\Src\Document\AbstractSegment.cs
Dis not use it due to the name.
The class is only used in this file.

mstv commented:
The first commit does not detect / skip all moved lines. This is fixed up by the second commit. But the first commit should correctly handle all cases.

> You have provided another example commit with commit https://github.com/gitextensions/gitextensions/commit/1e5d99ac23f47cce95ff7cbcbf0d1f3d963a053a, file src/app/GitUI/Editor/Diff/DiffHighlightService.cs, lines 210 / 180 ff.

first commit 4aed0619c1d41bc37c5462f97dabae7de07d6b31:

![image](https://github.com/user-attachments/assets/0765c8d6-25a4-4cae-a955-fdb45d96cd96)

second commit 30da54f51fa18b10b60db091433367279bd2ee90:

![image](https://github.com/user-attachments/assets/1d65d211-4f69-4073-bf9d-cf2c9da850fd)

gerhardol commented:
>     * test `CanGetAddedMovedDiffInfo` needs to be adapted

fixed (diff file was added as a displaycase for #119330 and I updated the file after your comments and missed to add this test to my playlist...)

> The first commit does not detect / skip all moved lines. This is fixed up by the second commit. But the first commit should correctly handle all cases.

I would rather squash the commits, there are too many changes that will not be used.

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -104,279 +107,313 @@ void SetIfUnsetInGit(string key, string value)
-            char r = getCharAt(lineRemoved.Offset + removedOffset);
-
-            if (a != r)
+        string getText(ISegment line)
```
```suggestion
        return;

        string GetText(ISegment line)
```

gerhardol replied:
usinga lambda instead

gerhardol replied:
will adopt some of this, for the tests

mstv reviewed (commented):
I would prefer to split the PR after the second commit.

The second commit should remove `ref found` from `GetBlockOfLines` already.
`maxLine` rather is `endLine`.

gerhardol commented on src/app/GitUI/Editor/Diff/RangeDiffHighlightService.cs:
```diff
@@ -46,16 +34,15 @@ public override void SetLineControl(DiffViewerLineNumberControl lineNumbersContr
+    public override bool IsSearchMatch(DiffViewerLineNumberControl lineNumbersControl, int indexInText)
+        => lineNumbersControl.GetLineInfo(indexInText)?.LineType is DiffLineType.Header;
+
+    public override string[] GetFullDiffPrefixes() => _diffFullPrefixes;
```
Do you mean the order of functions (unchanged, other code in between was changed) or the IsSearch call (FileViewer uses that)?

mstv replied:
> Do you mean the order of functions

Yes, (`GetGitCommandConfiguration` and) `IsSearchMatch` should have been / be moved behind `GetFullPrefixes`.

> unchanged

Only visible in BC after manually matching the lines.

gerhardol replied:
no change now

gerhardol commented:
> I would prefer to split the PR after the second commit.

Can do that after #11862 is merged.

gerhardol commented:
Rebased on https://github.com/gitextensions/gitextensions/pull/11862 dropped the "gap" part (except the refactoring) and the cache part.

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+    private void MarkInlineDifferences(IDocument document)
+    {
+        int index = 0;
+        List<DiffLineInfo> diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(i => i.LineNumInDiff)];
```
What does this syntax mean? `[.. orderedEnumerable]` in this single usage is a weird expression for `orderedEnumerable.ToArray()`. This ".." is named "spread element".
I prefer the explicit syntax although the Roslyn analyzer advertises the new syntax - not requesting a change.
```suggestion
        DiffLineInfo[] diffLines = _diffLinesInfo.DiffLines.Values.OrderBy(l => l.LineNumInDiff).ToArray();
```
but at least
```suggestion
        DiffLineInfo[] diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(l => l.LineNumInDiff)];
```

gerhardol replied:
spread operator, all from start to end. 
I do like the syntax, but as you want to change to an array I will change

mstv replied:
I will get used to that syntax. It can be used for array, too.
I.e. `ToArray()` is not necessary: My second suggestion in the initial comment works.

gerhardol replied:
applied the second suggestion

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
-        if (!_useGitColoring)
-        {
-            return;
+            foreach (TextMarker marker in GetDifferenceMarkers(GetText, linesRemoved, linesAdded, diffContentOffset))
+            {
+                markerStrategy.AddMarker(marker);
+            }
```
```suggestion
            markerStrategy.AddMarkers(GetDifferenceMarkers(GetText, linesRemoved, linesAdded, diffContentOffset));
```

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -368,43 +391,6 @@ private static (int LengthIdenticalAtStart, int LengthIdenticalAtEnd) AddDiffere
-        int line = 0;
-        bool found = false;
-
-        // Skip the first pair of removed / added lines, which uses to contain the filenames - but without highlighting
-        _ = GetRemovedLines(document, ref line, ref found);
-        if (found)
-        {
-            _ = GetAddedLines(document, ref line, ref found);
-        }
```
How / does the DiffLineNumAnalyzer exclude such lines?

gerhardol replied:
DiffLineType.Minus/DiffLineType.Plus is only set when a @@ header has been found

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
-            => invertMatch ^ LinePrefixHelper.DoesLineStartWith(document, offset, prefixStr);
-    }
+            // git-diff presents the removed lines directly followed by the added in a "block"
+            IReadOnlyList<ISegment> linesRemoved = GetBlockOfLines(diffLines, DiffLineType.Minus, ref index, ref found);
```
`found` is not evaluated.

gerhardol replied:
It passes information from minus to plus, if a new block is started.
Inline markers are only considered for minus directly followed by plus lines
This could be rewritten to remove found and continue if there are no linesRemoved

gerhardol replied:
I changed this. Note that the code increases in size (a few lines) and is not more efficient. But slightly easier to understand.

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+    {
+        int index = 0;
+        List<DiffLineInfo> diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(i => i.LineNumInDiff)];
+        bool found = false;
```
Used in the loop only. Declare there where it is set to `false` again.

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+        string GetText(ISegment line)
+        {
+            int len = line.Length - diffContentOffset;
+            if (line.Offset + line.Length >= document.TextLength)
+            {
+                // This is likely a test, where last line do not have a newline
+                --len;
+            }
```
Where does this come from?
I guess before this PR, `MarkInlineDifferences` did not see line feeds.
Perhaps, invert the condition?
Otherwise, `Math.Min` would be safer.

mstv replied:
I have guessed / remembered correctly. This PR _does_ change the behavior: Line feeds are fed to `AddDifferenceMarkers(List<TextMarker> markers, Func<ISegment, string> getText, ISegment lineRemoved, ISegment lineAdded, int beginOffset)` now.
The follow-up PR does not really change the behavior of this function. It mainly adapts to `string text` as replacement for `document`.

---

I do not understand neither what this condition shall achieve nor the comment. Where does it come from?

gerhardol replied:
There is a newline in patches that Git creates, the pesky \no newline.
Some test cases do not have the newline

mstv replied:
Why should we skip the last char if the newline is missing?!?

`line.Offset + line.Length == document.TextLength` is a valid situation. There is no need for adapting `len`.
`line.Offset + line.Length > document.TextLength` means `line` is kind of invalid regarding `document`. Use `Math.Min` in order to be safe because `--len` might not be enough.

In the follow-up it becomes:
```
        string GetText(ISegment line)
        {
            int len = line.Length - beginOffset;
            if (line.Offset + len == text.Length)
            {
                // This is likely a test, where last line do not have a newline
                --len;
            }

            return text.Substring(line.Offset + beginOffset, len);
        }
```
Same question.

gerhardol replied:
The core issue was in DiffLineNumAnalyzer.cs, could remove this fix

mstv replied:
> The core issue was in DiffLineNumAnalyzer.cs, could remove this fix

better
I took the liberty to push a proper support of CRLF.

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
     }
 
+    private List<ISegment> GetAllLines(DiffLineType diffLineType)
+    => _diffLinesInfo?.DiffLines.Where(i => i.Value.LineType == diffLineType && i.Value.LineSegment is not null)
+            .Select(i => i.Value.LineSegment)
```
```suggestion
    => _diffLinesInfo?.DiffLines.Where(l => l.Value.LineType == diffLineType && l.Value.LineSegment is not null)
            .Select(l => l.Value.LineSegment)
```

gerhardol replied:
I prefer to keep this, it is changed in #11981 - I would have preferred to keep this in one PR....

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+        }
     }
 
+    private List<ISegment> GetAllLines(DiffLineType diffLineType)
```
It is `private` and iterated once only, i.e. no need for `.ToList()`.
```suggestion
    private IEnumerable[] GetAllLines(DiffLineType diffLineType)
```

mstv commented on src/app/GitUI/Editor/Diff/DiffHighlightService.cs:
```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+            {
+                if (!found)
+                {
+                    // Start of block is not found yet.
```
```suggestion
                    // Start of block not found yet
```

mstv reviewed (approved):
👍

gerhardol commented:
Plan to squash and merge this in two days.
May separate the delimiters change. 
As the first commit is not enough, I will squash the commits (do not bother to switch the order, that would fix this too)
